Remote Web Inspector: [Cocoa] `inspectable` API

patrickangle opened the pull request:
#### 3d99c5a137548afb19025ccd16979e9357a8cde7
<pre>
Remote Web Inspector: [Cocoa] `inspectable` API
<a href="https://bugs.webkit.org/show_bug.cgi?id=245986">https://bugs.webkit.org/show_bug.cgi?id=245986</a>
rdar://100736163

Reviewed by Devin Rousso.

Add API to allow applications to explicitly mark a WKWebView or JSContext as inspectable. This flips assumptions about
inspectability of content. Previously the default state for inspectability was `true`, with a platform-specific deamon
making decisions about which applications would actually be inspectable on a system. This patch moves that decision into
WebKit/WebCore/JavaScriptCore instead, and flips the default for inspectability to be `false` unless an application
would have previously been always-inspectable. Those cases include:
- Applications with the `get-task-allow`/`com.apple.security.get-task-allow` development entitlements when linked
against an SDK without the API.
- Applications with the `com.apple.webinspector.allow`/`com.apple.private.webinspector.allow-remote-inspection`
entitlements.

These entitlements will no longer be necessary to enable inspection, and instead the new `inspectable` API should be
used.

* Source/JavaScriptCore/API/JSContext.h:
* Source/JavaScriptCore/API/JSContext.mm:
(-[JSContext inspectable]):
(-[JSContext setInspectable:]):
(-[JSContext _remoteInspectionEnabled]):
(-[JSContext _setRemoteInspectionEnabled:]):
* Source/JavaScriptCore/API/JSContextPrivate.h:
* Source/JavaScriptCore/API/JSContextRef.cpp:
(JSGlobalContextCreateInGroup):
(JSGlobalContextIsInspectable):
(JSGlobalContextSetInspectable):
(JSGlobalContextGetRemoteInspectionEnabled):
(JSGlobalContextSetRemoteInspectionEnabled):
* Source/JavaScriptCore/API/JSContextRef.h:
* Source/JavaScriptCore/API/JSContextRefPrivate.h:
* Source/JavaScriptCore/API/JSRemoteInspector.cpp:
(mainProcessHasEntitlement):
(defaultStateForRemoteInspectionEnabledByDefault):
(JSRemoteInspectorGetInspectionEnabledByDefault):
* Source/JavaScriptCore/API/JSRemoteInspector.h:
* Source/JavaScriptCore/API/WebKitAvailability.h:
* Source/JavaScriptCore/API/tests/testIncludes.m:
* Source/JavaScriptCore/config.h:
* Source/JavaScriptCore/inspector/JSGlobalObjectInspectorController.cpp:
(Inspector::JSGlobalObjectInspectorController::developerExtrasEnabled const):
* Source/JavaScriptCore/inspector/remote/RemoteInspectionTarget.cpp:
(Inspector::RemoteInspectionTarget::remoteControlAllowed const):
(Inspector::RemoteInspectionTarget::inspectable const):
(Inspector::RemoteInspectionTarget::setInspectable):
(Inspector::RemoteInspectionTarget::pauseWaitingForAutomaticInspection):
(Inspector::RemoteInspectionTarget::setRemoteDebuggingAllowed): Deleted.
* Source/JavaScriptCore/inspector/remote/RemoteInspectionTarget.h:
* Source/JavaScriptCore/inspector/remote/RemoteInspector.h:
* Source/JavaScriptCore/inspector/remote/RemoteInspectorConstants.h:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
(Inspector::globalStateForNotify):
(Inspector::RemoteInspector::singleton):
(Inspector::RemoteInspector::parentProcessAuditToken):
(Inspector::RemoteInspector::listingForInspectionTarget const):
(Inspector::globalAutomaticInspectionState): Deleted.
* Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorGlib.cpp:
(Inspector::RemoteInspector::listingForInspectionTarget const):
* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorSocket.cpp:
(Inspector::RemoteInspector::listingForInspectionTarget const):
* Source/JavaScriptCore/jsc.cpp:
(CommandLine::parseArguments):
(runJSC):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::setInspectable):
(JSC::JSGlobalObject::inspectable const):
(JSC::JSGlobalObject::setRemoteDebuggingEnabled): Deleted.
(JSC::JSGlobalObject::remoteDebuggingEnabled const): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
* Source/WebCore/config.h:
* Source/WebCore/inspector/InspectorController.cpp:
(WebCore::InspectorController::enabled const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::inspectable const):
(WebCore::Page::setInspectable):
(WebCore::Page::remoteInspectionAllowed const): Deleted.
(WebCore::Page::setRemoteInspectionAllowed): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::ServiceWorkerThreadProxy):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageGetAllowsRemoteInspection):
(WKPageSetAllowsRemoteInspection):
* Source/WebKit/UIProcess/API/Cocoa/APISerializedScriptValueCocoa.mm:
(API::SharedJSContext::ensureContext):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView inspectable]):
(-[WKWebView setInspectable:]):
(-[WKWebView _allowsRemoteInspection]):
(-[WKWebView _setAllowsRemoteInspection:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Inspector/mac/WKInspectorViewController.mm:
(-[WKInspectorViewController webView]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::m_fullscreenVideoTextRecognitionTimer):
(WebKit::WebPageProxy::inspectable const):
(WebKit::WebPageProxy::setInspectable):
(WebKit::WebPageProxy::allowsRemoteInspection const): Deleted.
(WebKit::WebPageProxy::setAllowsRemoteInspection): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit2Prefix.h:
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):
(-[WebView allowsRemoteInspection]):
(-[WebView setAllowsRemoteInspection:]):
* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController awakeFromNib]):
* Tools/MobileMiniBrowser/MobileMiniBrowserFramework/WebViewController.m:
(-[WebViewController createWebView]):
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView initWithFrame:configuration:]):

Canonical link: <a href="https://commits.webkit.org/255557@main">https://commits.webkit.org/255557@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80c553ea91c1b5f8a8cc3775a54ea6f92a53368c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92904 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2112 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23486 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102627 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96906 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2117 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30450 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85299 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/98757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98569 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79387 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28362 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/83364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71479 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/84238 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36850 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16990 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/79303 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34653 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/18180 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27474 "Failed to checkout and rebase branch from PR 5259") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/3848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38523 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/40780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/81932 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1774 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40443 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/37360 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18534 "Found 1 new JSC stress test failure: stress/call-apply-exponential-bytecode-size.js.mini-mode (failure)") | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/JavaScriptCore/inspector/remote/RemoteInspectorConstants.h:
```diff
@@ -36,10 +36,14 @@
+// COMPATIBILITY(macOS 13): The key string is intentionally mismatched to support old relays.
+#define WIRGlobalNotifyStateName                    "com.apple.webinspectord.automatic_inspection_enabled"
+#define WIRGlobalNotifyStateAutomaticInspection     1 << 0
+#define WIRGlobalNotifyStateSimulateCustomerInstall 1 << 1
```
I think maybe we should make this the last bit (i.e. `1 << 63`) so that if we need to add more "real" things later on we don't have a "hole" (especially if this is removed in the future).

dcrousso commented on Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
```diff
@@ -72,15 +72,17 @@ static bool canAccessWebInspectorMachPort()
-    notify_get_state(token, &automaticInspectionEnabled);
-    return automaticInspectionEnabled == 1;
+    m_automaticInspectionEnabled = state & WIRGlobalNotifyStateAutomaticInspection;
+    m_simulateCustomerInstall = state & WIRGlobalNotifyStateSimulateCustomerInstall;
```
I think we should also put this behind `allowInternalSecurityPolicies` (and make sure not to even send it inside `webinspectord` using a similar check) so that we don't even bother with this logic when not dealing with a development system.

patrickangle committed:
Remote Web Inspector: [Cocoa] `inspectable` API
https://bugs.webkit.org/show_bug.cgi?id=245986
rdar://100736163

Reviewed by Devin Rousso.

Add API to allow applications to explicitly mark a WKWebView or JSContext as inspectable. This flips assumptions about
inspectability of content. Previously the default state for inspectability was `true`, with a platform-specific deamon
making decisions about which applications would actually be inspectable on a system. This patch moves that decision into
WebKit/WebCore/JavaScriptCore instead, and flips the default for inspectability to be `false` unless an application
would have previously been always-inspectable. Those cases include:
- Applications with the `get-task-allow`/`com.apple.security.get-task-allow` development entitlements when linked
against an SDK without the API.
- Applications with the `com.apple.webinspector.allow`/`com.apple.private.webinspector.allow-remote-inspection`
entitlements.

These entitlements will no longer be necessary to enable inspection, and instead the new `inspectable` API should be
used.

* Source/JavaScriptCore/API/JSContext.h:
* Source/JavaScriptCore/API/JSContext.mm:
(-[JSContext inspectable]):
(-[JSContext setInspectable:]):
(-[JSContext _remoteInspectionEnabled]):
(-[JSContext _setRemoteInspectionEnabled:]):
* Source/JavaScriptCore/API/JSContextPrivate.h:
* Source/JavaScriptCore/API/JSContextRef.cpp:
(JSGlobalContextCreateInGroup):
(JSGlobalContextIsInspectable):
(JSGlobalContextSetInspectable):
(JSGlobalContextGetRemoteInspectionEnabled):
(JSGlobalContextSetRemoteInspectionEnabled):
* Source/JavaScriptCore/API/JSContextRef.h:
* Source/JavaScriptCore/API/JSContextRefPrivate.h:
* Source/JavaScriptCore/API/JSRemoteInspector.cpp:
(mainProcessHasEntitlement):
(defaultStateForRemoteInspectionEnabledByDefault):
(JSRemoteInspectorGetInspectionEnabledByDefault):
* Source/JavaScriptCore/API/JSRemoteInspector.h:
* Source/JavaScriptCore/API/WebKitAvailability.h:
* Source/JavaScriptCore/API/tests/testIncludes.m:
* Source/JavaScriptCore/config.h:
* Source/JavaScriptCore/inspector/JSGlobalObjectInspectorController.cpp:
(Inspector::JSGlobalObjectInspectorController::developerExtrasEnabled const):
* Source/JavaScriptCore/inspector/remote/RemoteInspectionTarget.cpp:
(Inspector::RemoteInspectionTarget::remoteControlAllowed const):
(Inspector::RemoteInspectionTarget::inspectable const):
(Inspector::RemoteInspectionTarget::setInspectable):
(Inspector::RemoteInspectionTarget::pauseWaitingForAutomaticInspection):
(Inspector::RemoteInspectionTarget::setRemoteDebuggingAllowed): Deleted.
* Source/JavaScriptCore/inspector/remote/RemoteInspectionTarget.h:
* Source/JavaScriptCore/inspector/remote/RemoteInspector.h:
* Source/JavaScriptCore/inspector/remote/RemoteInspectorConstants.h:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
(Inspector::globalStateForNotify):
(Inspector::RemoteInspector::singleton):
(Inspector::RemoteInspector::parentProcessAuditToken):
(Inspector::RemoteInspector::listingForInspectionTarget const):
(Inspector::globalAutomaticInspectionState): Deleted.
* Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorGlib.cpp:
(Inspector::RemoteInspector::listingForInspectionTarget const):
* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorSocket.cpp:
(Inspector::RemoteInspector::listingForInspectionTarget const):
* Source/JavaScriptCore/jsc.cpp:
(CommandLine::parseArguments):
(runJSC):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::setInspectable):
(JSC::JSGlobalObject::inspectable const):
(JSC::JSGlobalObject::setRemoteDebuggingEnabled): Deleted.
(JSC::JSGlobalObject::remoteDebuggingEnabled const): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
* Source/WebCore/config.h:
* Source/WebCore/inspector/InspectorController.cpp:
(WebCore::InspectorController::enabled const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::inspectable const):
(WebCore::Page::setInspectable):
(WebCore::Page::remoteInspectionAllowed const): Deleted.
(WebCore::Page::setRemoteInspectionAllowed): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::ServiceWorkerThreadProxy):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageGetAllowsRemoteInspection):
(WKPageSetAllowsRemoteInspection):
* Source/WebKit/UIProcess/API/Cocoa/APISerializedScriptValueCocoa.mm:
(API::SharedJSContext::ensureContext):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView inspectable]):
(-[WKWebView setInspectable:]):
(-[WKWebView _allowsRemoteInspection]):
(-[WKWebView _setAllowsRemoteInspection:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Inspector/mac/WKInspectorViewController.mm:
(-[WKInspectorViewController webView]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::m_fullscreenVideoTextRecognitionTimer):
(WebKit::WebPageProxy::inspectable const):
(WebKit::WebPageProxy::setInspectable):
(WebKit::WebPageProxy::allowsRemoteInspection const): Deleted.
(WebKit::WebPageProxy::setAllowsRemoteInspection): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit2Prefix.h:
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):
(-[WebView allowsRemoteInspection]):
(-[WebView setAllowsRemoteInspection:]):
* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController awakeFromNib]):
* Tools/MobileMiniBrowser/MobileMiniBrowserFramework/WebViewController.m:
(-[WebViewController createWebView]):
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView initWithFrame:configuration:]):

Canonical link: https://commits.webkit.org/255557@main
